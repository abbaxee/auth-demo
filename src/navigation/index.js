import React, {useEffect} from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
import Login from '../screens/Login';
import Register from '../screens/Register';
import UserData from '../screens/UserData';
import {StyleSheet} from 'react-native';
import auth from '@react-native-firebase/auth';

import AsyncStorage from '@react-native-community/async-storage';

export const AuthContext = React.createContext({});

const reducer = (state, action) => {
  switch (action.type) {
    case 'LOGIN':
      return {
        ...state,
        user: action.payload,
      };
    case 'LOGOUT':
      return {
        ...state,
        user: null,
      };
    default:
      return state;
  }
};

const AppNavigator = () => {
  const {Screen, Navigator} = createStackNavigator();
  const initialState = {user: null};
  const [authState, dispatch] = React.useReducer(reducer, initialState);

  useEffect(() => {
    const initializeAuth = async () => {
      const user = await AsyncStorage.getItem('@userData');
      if (user) {
        dispatch({type: 'LOGIN', payload: JSON.parse(user)});
      }
    };
    initializeAuth();
  }, []);

  const logout = React.useCallback(() => {
    AsyncStorage.removeItem('@userData');
    dispatch({type: 'LOGOUT'});
  }, [dispatch]);

  return (
    <AuthContext.Provider value={{state: authState, dispatch, logout}}>
      <NavigationContainer>
        <Navigator screenOptions={{headerTitleStyle: styles.headerTitleStyle}}>
          {!authState.user ? (
            <>
              <Screen name="Login" component={Login} />
              <Screen name="Register">
                {(props) => <Register auth={auth} {...props} />}
              </Screen>
            </>
          ) : (
            <Screen
              name="User"
              component={UserData}
              options={{title: 'User Data'}}
            />
          )}
        </Navigator>
      </NavigationContainer>
    </AuthContext.Provider>
  );
};

const styles = StyleSheet.create({
  headerTitleStyle: {
    fontWeight: 'bold',
    fontSize: 24,
  },
});

export default AppNavigator;
