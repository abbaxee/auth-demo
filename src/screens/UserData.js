import React from 'react';
import {SafeAreaView, View, Text, StyleSheet, Image} from 'react-native';
import Button from '../components/Button';
import {AuthContext} from '../navigation';
import {GREY_THREE, WHITE_COLOR} from '../styles/colors';
import container from '../styles/container';
// import AsyncStorage from '@react-native-community/async-storage';

const DetailText = ({title, text}) => (
  <Text style={styles.detailText}>
    <Text style={styles.titleText}>{title}: </Text> {text}
  </Text>
);

const UserData = () => {
  const {state, logout} = React.useContext(AuthContext);

  const {user} = state;

  return (
    <>
      <SafeAreaView>
        <View style={container.medium}>
          <View style={styles.content}>
            <View>
              <Image
                style={styles.avatar}
                resizeMode="contain"
                source={{uri: 'https://picsum.photos/200'}}
              />
            </View>

            <View style={styles.detail}>
              <DetailText title="Name" text={user && user.displayName} />
              <DetailText title="Email" text={user && user.email} />
            </View>
          </View>
          <Button text="Logout" onPress={logout} />
        </View>
      </SafeAreaView>
    </>
  );
};

const styles = StyleSheet.create({
  content: {
    // alignItems: 'center',
    marginBottom: 20,
    backgroundColor: WHITE_COLOR,
    padding: 40,
    borderRadius: 6,
  },
  avatar: {
    width: 140,
    height: 140,
    borderRadius: 70,
  },
  detail: {
    marginTop: 20,
  },
  detailText: {
    fontSize: 14,
    paddingVertical: 2,
    color: GREY_THREE,
  },
  titleText: {
    fontWeight: 'bold',
  },
});

export default UserData;
