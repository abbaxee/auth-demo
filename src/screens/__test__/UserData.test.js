import React from 'react';
import {render} from '@testing-library/react-native';
import UserData from '../UserData';
import {AuthContext} from '../../navigation';

test('User Data renders properly', () => {
  const state = {user: {displayName: 'John Doe', email: 'john@doe.com'}};
  const logout = jest.fn();

  const instance = render(
    <AuthContext.Provider value={{state, logout}}>
      <UserData />
    </AuthContext.Provider>,
  ).toJSON();
  expect(instance).toMatchSnapshot();
});
