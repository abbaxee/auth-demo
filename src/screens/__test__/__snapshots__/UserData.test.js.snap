// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Data renders properly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "padding": 24,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": 6,
          "marginBottom": 20,
          "padding": 40,
        }
      }
    >
      <View>
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://picsum.photos/200",
            }
          }
          style={
            Object {
              "borderRadius": 70,
              "height": 140,
              "width": 140,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#333333",
              "fontSize": 14,
              "paddingVertical": 2,
            }
          }
        >
          <Text
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Name
            : 
          </Text>
           
          John Doe
        </Text>
        <Text
          style={
            Object {
              "color": "#333333",
              "fontSize": 14,
              "paddingVertical": 2,
            }
          }
        >
          <Text
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Email
            : 
          </Text>
           
          john@doe.com
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#333333",
          "borderRadius": 6,
          "marginVertical": 20,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 18,
            "fontWeight": "bold",
            "textTransform": "uppercase",
          }
        }
      >
        Logout
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`User Data renders properly: 
    <RCTSafeAreaView
      emulateUnlessSupported={true}
    >
      <View
        style={
          Object {
            "padding": 24,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#ffffff",
              "borderRadius": 6,
              "marginBottom": 20,
              "padding": 40,
            }
          }
        >
          <View>
            <Image
              resizeMode="contain"
              source={
                Object {
                  "uri": "https://picsum.photos/200",
                }
              }
              style={
                Object {
                  "borderRadius": 70,
                  "height": 140,
                  "width": 140,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "marginTop": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#333333",
                  "fontSize": 14,
                  "paddingVertical": 2,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontWeight": "bold",
                  }
                }
              >
                Name
                :
              </Text>

              John Doe
            </Text>
            <Text
              style={
                Object {
                  "color": "#333333",
                  "fontSize": 14,
                  "paddingVertical": 2,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontWeight": "bold",
                  }
                }
              >
                Email
                :
              </Text>

              john@doe.com
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#333333",
              "borderRadius": 6,
              "marginVertical": 20,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 18,
                "fontWeight": "bold",
                "textTransform": "uppercase",
              }
            }
          >
            Logout
          </Text>
        </View>
      </View>
    </RCTSafeAreaView>
   1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "padding": 24,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": 6,
          "marginBottom": 20,
          "padding": 40,
        }
      }
    >
      <View>
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://picsum.photos/200",
            }
          }
          style={
            Object {
              "borderRadius": 70,
              "height": 140,
              "width": 140,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#333333",
              "fontSize": 14,
              "paddingVertical": 2,
            }
          }
        >
          <Text
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Name
            : 
          </Text>
           
          John Doe
        </Text>
        <Text
          style={
            Object {
              "color": "#333333",
              "fontSize": 14,
              "paddingVertical": 2,
            }
          }
        >
          <Text
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Email
            : 
          </Text>
           
          john@doe.com
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#333333",
          "borderRadius": 6,
          "marginVertical": 20,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 18,
            "fontWeight": "bold",
            "textTransform": "uppercase",
          }
        }
      >
        Logout
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
